packages/query-core/src/queryClient.ts

```diff
@@ -106,11 +107,22 @@ export class QueryClient {
     return this.mutationCache.findAll({ ...filters, fetching: true }).length
   }
 
+  getQueryData<TData = unknown, Filters extends QueryFilters = QueryFilters>(
+    filters: Filters,
+  ): TData | undefined
   getQueryData<TData = unknown>(
     queryKey: QueryKey,
     filters?: QueryFilters,
+  ): TData | undefined
+  getQueryData<TData = unknown, Filters extends QueryFilters = QueryFilters>(
```

Comment:
the generic trick is nice, but it sadly won't work if we provide a generic to TData:

```
const query = {
  queryKey,
  queryFn
}
queryClient.getQueryData<string>(query)
```

this will still fail because TypeScript has no partial inference for generics, so because we provide the first generic, the second one will default to `QueryFilters`, and then TS complains that we cannot pass the `queryFn`...
and in order to work with `getQueryData` in a meaningful way in TypeScript, it is necessary to supply the `TData` generic.

Reply:
Indeed.
So another option would be the solution of my first commit ?

Reply:
One cool thing about using the extended filters, is that if your factory returns has a `queryFn` the return type is now inferred : 
```ts
const query = {
  queryKey,
  queryFn // return a string
}
const data = queryClient.getQueryData(query);
//        ^? typeof data : string
const data = queryClient.getQueryData<string>(query); // stil works
const data = queryClient.getQueryData<number>(query); // typescript error number and queryFn missmatch
```